Add PlaceAlternateBuilding trait

pchote opened the pull request:
This PR allows `Building`s to define an alternate version that can be selected and placed by holding Shift while placing the actor. This adds another option to the toolbox for the next time discussion about north-south imbalance flares up, and i'm sure modders will be able to find some more creative usecases for it.

This uses separate actors because we can't (yet) modify the building previews or footprint using conditions.  We can consider changing this to use conditions in the future once these have been fixed.

I've added this to the TS gates, and added a temporary testcase that flips the TS refinery - demonstrating "silent" alternates that aren't buildable on their own. The testcase isn't useful as a real feature because the x-flip casts the shadow in the wrong direction and doesn't significantly help with the balance issue.

pchote committed:
Add PlaceBuildingVariant trait.

MustaphaTR commented:
I'm a bit unsure about the holding shift part. Could be more useful in terms of modding if you could switch between 2+ actors by clicking shift. Ctrl (or CTRL+SHIFT) could be used for going thru the list other way around.

pchote commented:
We already use shift to modify the behaviour for walls, so I figured it made sense to apply that here too.

MustaphaTR commented:
Hmm, i wasn't aware of that. But still allowing multipile sounds better than allowing just 2 to me.

pchote commented:
Shift and Ctrl are modifier keys - it would be bad UX and bad code (they can't be detected via normal button presses, so would need to be bodged by observing modifier state changes from Tick) to implement them for state changes. Using the mousewheel was raised as an option in Discord, but this is also bad UX (scrolling is a continuous action, and should not be used to toggle discrete states) and makes the feature difficult to access for players using trackpads or other input methods without a dedicated scrollwheel.

The only option that IMO would be reasonable to toggle discrete states would be the tab key, at the expense of disabling queue toggling while the OG is active.

In any case, changing this from a modifier-based alternative to toggling between N discrete states would require a complete rewrite of the code here, so if there is a consensus that this is the way forward I will close this and leave the reimplementation to somebody else.

pchote commented:
I plan to file a prereq PR that reworks the building preview rendering before coming back to this.

pchote commented:
Updated to be based on #16553, but the changes to support more than one variant are not yet completed.

matjaeck commented on OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs:
```diff
@@ -172,89 +231,58 @@ bool AcceptsPlug(CPos cell, PlugInfo plug)
+			var plugInfo = activeVariant.PlugInfo;
+			var lineBuildInfo = activeVariant.LineBuildInfo;
+			var preview = activeVariant.Preview;
 
```
`Orders/PlaceBuildingOrderGenerator.cs(242,1): error SA1507: Code should not contain multiple blank lines in a row`

pchote committed:
Add key handling to order generators.

pchote commented:
Updated to support toggling between an arbitrary number of variants using a mod-definable hotkey.

It doesn't make sense to expose a dedicated hotkey *just* for rotating gates in TS, so I have removed that ability from the PR.  This feature will now just be available for modders instead of being used in the default mods.

pchote commented:
rebased.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Buildings/PlaceBuildingVariant.cs:
```diff
@@ -0,0 +1,28 @@
+		[FieldLoader.Require]
+		[ActorReference(typeof(BuildingInfo))]
+		[Desc("Alternate building actor to place.")]
+		public readonly string Actor = null;
```
Imho that would be better implemented as list of actors.

pchote replied:
Ok, done.

pchote commented:
Updated.

pchote commented:
Testcase removed and rebased.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/538a12f50da5e00f630ede21944379e1cf047c96...7bc113d8f671efffcc0186e38b50f9f4c18ec123)

GraionDilach commented on OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs:
```diff
@@ -94,29 +123,34 @@ PlaceBuildingCellType MakeCellType(bool valid, bool lineBuild = false)
 		{
-			if (mi.Button == MouseButton.Right)
-				world.CancelInputMode();
+			if ((mi.Button == MouseButton.Left && mi.Event == MouseInputEvent.Down) || (mi.Button == MouseButton.Right && mi.Event == MouseInputEvent.Up))
```
The AI cannot click to pass this check, which have led to #16679.
